### DISCUSSION TITLE
Feature Request: Allow using multiple swagger UI CSS

### DISCUSSION THREAD

**Discussion by ajatkj** [fastapi/fastapi#10927] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Proposed change in fastapi/openapi/docs.py

def get_swagger_ui_html(..., 
swagger_css_url: Annotated[
        union[str, list[str]],
        Doc(
            """
            The URL to use to load the Swagger UI CSS.

            It is normally set to a CDN URL.
            """
        ),
    ] = "https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.9.0/swagger-ui.css",
...)

and use this list in generating the html code.
```

### Description

To have a better flexibility on swagger UI CSS configuration, I propose being able to use multiple CSS files in `get_swagger_ui_html` function. i.e. current string variable will work as is (so backward compatible) and add a union with `list[str]` and add multiple `<Link rel...>` lines in the generated html.

Use case:
I can have a base CSS with a default set of CSS custom properties (defining various colors, svgs backgrounds and so on) and additional CSS files which can override those properties (using !important) to give more theming options.

### Operating System

macOS

### FastAPI Version

0.108.0

### Pydantic Version

2.5.3

### Python Version

3.12.1

**Comment by hasansezertasan** [comment]
Did you try overriding this feature or do you have an example?

**Reply by ajatkj** [reply]
I have tried it locally and it works as expected. Example of what?

**Reply by hasansezertasan** [reply]
Example code of the implementation.

**Reply by ajatkj** [reply]
Sure. Below is the relevant code -

`fastapi/openapi/docs.py`
```py3
swagger_css_url: Annotated[
    Union[str, list[str]],
    Doc(
        """
        The URL to use to load the Swagger UI CSS.

        It is normally set to a CDN URL.
        """
    ),
] = "https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.9.0/swagger-ui.css",
```

```py3
if isinstance(swagger_css_url, list):
    html+= "\n".join([f"""<link type="text/css" rel="stylesheet" href="{url}">""" for url in swagger_css_url])
else:
    html+= f"""<link type="text/css" rel="stylesheet" href="{swagger_css_url}">"""
```

Function call
```py3
async def custom_swagger_ui(app: "FastAPI", prefix: str):
    return get_swagger_ui_html(
        openapi_url=f"{prefix}/openapi.json",
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/5.10.5/swagger-ui-bundle.js",
        swagger_css_url=[
            "https://cdn.jsdelivr.net/gh/ajatkj/swagger-ui-improved-theme/css/swagger-ui-improved.css",
            "/static/dark-theme.css",
        ],
    )
```

`/static/dark-theme.css`
```css
html {
  background: rgb(39, 30, 86) !important;
}
```

Generated html
```html
<!DOCTYPE html>
    <html>
    <head>
    <link type="text/css" rel="stylesheet" href="https://cdn.jsdelivr.net/gh/ajatkj/swagger-ui-improved-theme/css/swagger-ui-improved.css">
    <link type="text/css" rel="stylesheet" href="/static/dark-theme.css">
    <link rel="shortcut icon" href="https://fastapi.tiangolo.com/img/favicon.png">
    <title>The API - Swagger UI</title>
    </head>
    <body>
    <div id="swagger-ui">
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/5.10.5/swagger-ui-bundle.js"></script>
    <!-- `SwaggerUIBundle` is now available on the page -->
    <script>
    const ui = SwaggerUIBundle({
        url: '/user/openapi.json',
        "dom_id": "#swagger-ui",
        "layout": "BaseLayout",
        "deepLinking": true,
        "showExtensions": true,
        "showCommonExtensions": true,
        oauth2RedirectUrl: window.location.origin + '/docs/oauth2-redirect',
        presets: [
            SwaggerUIBundle.presets.apis,
            SwaggerUIBundle.SwaggerUIStandalonePreset
        ],
    })
    </script>
    </body>
    </html>
```

Generated doc-
![image](https://github.com/tiangolo/fastapi/assets/8344288/211560cb-dfa6-40b1-881e-4e4ceb29c5cd)

**Reply by hasansezertasan** [reply]
It looks quite good 👏. Extra JS and Extra CSS seem cool.

I upvoted the discussion. 👍

**Reply by ajatkj** [reply]
Thanks. Does this mean I can go ahead with the PR?

**Reply by hasansezertasan** [reply]
Oh, that's up to you. Who am I to tell you to go ahead or not? But I encourage you to give it a try.

**Comment by tiangolo** [comment, marked as answer]
Thanks! But I prefer to show people how to fully customize their docs UI any way they want as in here: https://fastapi.tiangolo.com/how-to/custom-docs-ui-assets/#self-hosting-javascript-and-css-for-docs

If someone, for example you, wanted to have a more customized docs UI, it's possible to entirely override the whole HTML, which is more flexible than just the list of CSS and JS files to include.

Given that, I'll pass on this request, but thanks for the interest! ☕

**Reply by ajatkj** [reply]
Thanks for the reply. I understand your decision but let me try to explain my reason behind this (not trying to make you reconsider your decision😜)
- the swagger js and css are very complicated. Sometimes you just want to find one element and change it. Don't need to host entire js/css. 
- if the underlying js/css is updated, I will have to retrofit everything I did on the newer version. 
- and of course what I mentioned in the discussion about being able to create themes by having one base css and other theming css just overriding custom properties. 

Again, not trying to push you, just trying to make a case. 

I am big a advocate of this framework btw.

**Comment by htran-ubed** [comment]
I created a workaround to include a custom css-file.

You need to get the default value for the swagger_css_url parameter and then add some html to it. Here's my working example:
```
import inspect

SWAGGER_UI_CUSTOM_CSS_URL = './static/css/swagger-ui.css'

def get_parameter_default_value(function, parameter):
    return inspect.signature(function).parameters.get(parameter).default

@app.get('/docs', include_in_schema = False)
def overridden_swagger():
    swagger_css_url = get_parameter_default_value(get_swagger_ui_html, 'swagger_css_url')

    return get_swagger_ui_html(
        swagger_css_url = f'{swagger_css_url}"><link type="text/css" rel="stylesheet" href="{SWAGGER_UI_CUSTOM_CSS_URL}',
    )
```
